server/src/main/java/io/deephaven/server/barrage/BarrageMessageProducer.java

```diff
@@ -974,139 +1010,278 @@ private void updateSubscriptionsSnapshotAndPropagate() {
-                }
 
-                postSnapshotViewport = haveViewport ? postSnapshotViewportBuilder.build() : null;
-                postSnapshotReverseViewport = haveViewport ? postSnapshotReverseViewportBuilder.build() : null;
+                    subscription.isFirstSnapshot = false;
 
-                if (!needsSnapshot) {
-                    // i.e. We have only removed subscriptions; we can update this state immediately.
-                    promoteSnapshotToActive();
+                    // get the current set for this viewport direction
+                    final WritableRowSet currentSet =
+                            subscription.targetReverseViewport ? reverseSnapshotRowSet : snapshotRowSet;
+
+                    // get the rows that we need that are already in the snapshot
+                    subscription.growingIncrementalViewport = subscription.growingRemainingViewport.extract(currentSet);
+                    if (rowsRemaining > 0) {
+                        try (final WritableRowSet additional = subscription.growingRemainingViewport.copy()) {
+
+                            // shrink the set of new rows to <= `rowsRemaining` size
+                            if (additional.size() > rowsRemaining) {
+                                final long key = additional.get(rowsRemaining);
+                                additional.removeRange(key, Long.MAX_VALUE - 1);
+
+                                // update the rows remaining
+                                subscription.growingRemainingViewport.removeRange(0, key - 1);
+                            } else {
+                                // all rows are satisfied
+                                subscription.growingRemainingViewport.clear();
+                            }
+
+                            // store the rowset that applies for this exact snapshot
+                            subscription.growingIncrementalViewport.insert(additional);
+
+                            // add the new rows to the upcoming snapshot
+                            currentSet.insert(additional);
+
+                            if (subscription.targetReverseViewport) {
+                                // add this set to the global reverse viewport (for scoping)
+                                reverseViewportBuilder.addRowSet(additional);
+                            } else {
+                                // add this set to the global forward viewport (for scoping)
+                                viewportBuilder.addRowSet(additional);
+                            }
+
+                            // decrement the remaining row count
+                            rowsRemaining -= additional.size();
+                        }
+                    }
+
+                    // "grow" the snapshot viewport with rows from the current snapshot
+                    if (viewportValid && subscription.viewport != null) {
```

Comment:
I believe this if block is now unnecessary per the `isFirstSnapshot` block above.